distribution/push_v2.go

```diff
@@ -207,39 +218,87 @@ func (pd *v2PushDescriptor) DiffID() layer.DiffID {
 	return pd.layer.DiffID()
 }
 
-func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.Output) (digest.Digest, error) {
+func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.Output) error {
 	diffID := pd.DiffID()
 
-	logrus.Debugf("Pushing layer: %s", diffID)
+	pd.pushState.Lock()
+	if _, ok := pd.pushState.remoteLayers[diffID]; ok {
+		// it is already known that the push is not needed and
+		// therefore doing a stat is unnecessary
+		pd.pushState.Unlock()
+		progress.Update(progressOutput, pd.ID(), "Layer already exists")
+		return nil
+	}
+	pd.pushState.Unlock()
 
 	// Do we have any blobsums associated with this layer's DiffID?
 	possibleBlobsums, err := pd.blobSumService.GetBlobSums(diffID)
 	if err == nil {
-		dgst, exists, err := blobSumAlreadyExists(ctx, possibleBlobsums, pd.repo, pd.layersPushed)
+		possibleDigests := make([]digest.Digest, len(possibleBlobsums))
+		for i := range possibleBlobsums {
+			possibleDigests[i] = possibleBlobsums[i].Digest
+		}
+		descriptor, exists, err := blobSumAlreadyExists(ctx, possibleBlobsums, pd.repo, pd.pushState)
```

Comment:
I think it makes sense to limit these checks to entries where `SourceRepository` matches the repository we're pushing to; otherwise we brute-force `Stat`s of every blobsum we know about before trying a cross-repository push.

I suggest changing `blobSumAlreadyExists` to take `[]metadata.BlobSum` as its argument, and doing this check inside `blobSumAlreadyExists`, to avoid the conversion to `[]digest.Digest`.

Note that we should also stat items where `SourceRepository` is empty, since checksums migrated from earlier versions won't have a `SourceRepository` value.